### ISSUE TITLE
[ASR] Bad results on 8000hz custom dataset

### ISSUE THREAD

**Issue by tudor-mare** [espnet/espnet#3013]
Hi,

I am currently trying to train an ASR solution based on the AN4 recipe. My dataset consists of 50hour transcribed audios sampled at 8000hz.

I tried using a small part of my dataset (1h, 1100 samples) so I experiment here and then move on the big dataset so I don't waste time, but I have problems training even on this small one.
As preprocessing I am applying a volume threshold at the start and at the end of the audios. I also augment the speed of the audios with 0.8,1.1,1.2 ratios.

Audio config:

> --use-energy=false   # only non-default option.
--sample-frequency=8000
--num-mel-bins=80    # similar to Google's setup.
// --num-ceps=80
--frame-length=25

The transcripts contain have a mean len of 30 characters. And the audio samples have a length of around 2 seconds.
I am training a char based LM with this config:

> layer: 2         # 2 for character LMs
unit: 650       # 650 for character LMs
opt: adam         # adam for character LMs
sortagrad: 0 # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
batchsize: 256    # 1024 for character LMs
epoch: 100       # number of epochs
patience: 0
maxlen: 150        # 150 for character LMs
dropout-rate: 0.5
gradclip: 0.5

![stats train LM](https://user-images.githubusercontent.com/34603830/109007126-77247180-76b4-11eb-94e1-3a594f2b2a06.png)

For my CTC/ATT I use this config:

>--minibatch related
batch-size: 32
maxlen-in: 300  # if input length  > maxlen_in, batchsize is automatically reduced
maxlen-out: 150 # if output length > maxlen_out, batchsize is automatically reduced
optimization related
sortagrad: 0 # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
opt: adadelta
epochs: 45
patience: 0
--scheduled sampling option
sampling-probability: 0.0
--encoder related
etype: vgglstm     # encoder architecture type
elayers: 2
eunits: 128
eprojs: 128
subsample: "1_2_2_1_1" # skip every n frame from input to nth layers
--decoder related
dlayers: 1
dunits: 128
--attention related
atype: location
adim: 128
aconv-chans: 5
aconv-filts: 32
--hybrid CTC/attention
mtlalpha: 0.5
grad-clip: 0.5

I use a small batchsize because of the size of my dataset. 

![statsVal no LM](https://user-images.githubusercontent.com/34603830/109007198-8d323200-76b4-11eb-9a74-bc38f805774e.png)

My decode config is as follows:

>--decoding parameter
beam-size: 4
penalty: 0.0
maxlenratio: 0.0
minlenratio: 0.0
ctc-weight: 0.3
lm-weight: 0.5

Results without LM:
|  SPKR    |  # Snt    # Wrd  |  Corr      Sub       Del      Ins      Err    S.Err  |
|  Sum/Avg |   229     14862  |  54.0     22.4      23.6     24.3     70.3    100.0  |

Results with LM:
|  SPKR    |  # Snt    # Wrd  |  Corr      Sub       Del      Ins      Err    S.Err  |
|  Sum/Avg |   229     14862  |  53.5     24.1      22.4     33.7     80.2    100.0  |

What I've tried:

- using the full dataset (50h) with more adim,aconv-filts,aconv-chan,eunits,edunits but after 1 day of training the loss looks very spiky and the output is full of < space > or it outputs same 'nema..lului' word that contains the most common prefix and suffix
- using 2.500.000 samples plus the transcriptions for LM model but the performance doesn't improve; it only prints <blank>< blank >...
- changed to a transformer model

I don't know if I should leave the model to train for longer, because I am afraid to waste a lot of time and then see that the model is not strong enough.

Any advice is greatly appreciated :) Thanks

**Comment by tudor-mare** [comment]
Update: I trained on full dataset with the same config as above and I get this results at the moment of writing (it still has 12 more hours to train).
![image](https://user-images.githubusercontent.com/34603830/109139156-dd1c0200-7763-11eb-8381-e7cd642a9006.png)

From what I see above I deduce that the ATT has not enough power to learn and the same for the LM model.
At the same time I find it weird that the cer_ctc is going down while the loss_ctc is more or less on a spiky plateau.
What hyperparameters should I change so the model has enough power to learn?

 Thanks

**Comment by sw005320** [comment]
First, I don't recommend you to test your model with only one hour of training data. We cannot fully train the attention model.
I recommend you use around 15 hours of training data at least.
The voxforge Italian recipe is a similar range and also gives us a reasonable performance.
After you confirm it, you can just increase the training data.

But if I am in the same situation, I can directly use 50 hours of your data. 
Smaller data just makes training more difficult.

Also, if you tune the hyper-parameters, I recommend you to tune optimization hyperparameters, e.g., learning rate, but our config is well-tuned in each recipe.

**Comment by tudor-mare** [comment]
Thanks for your response!

I have trained a model on the voxforge recipe (I used same config files) and have the following results:

![train](https://user-images.githubusercontent.com/34603830/109789063-996c4100-7c18-11eb-966f-e2e89f8f7b84.png)

![test](https://user-images.githubusercontent.com/34603830/109789078-a12be580-7c18-11eb-96be-61a64b7deb0f.png)

I can see that the Attention part of the model works as the accuracy on test is proportionate to the ATT loss. The model has problems on the CTC part. Is there a parameter that I should set according to the length of my samples/transcripts? 

From what I see in the decoding output the model outputs same slightly modified word 'nemai..reci'

> id: (002-002_00_0001)
Scores: (#C #S #D #I) 2 5 1 1
REF:  S T ă N C i L e * 
HYP:  * N E  M A i R e I 
Eval: D S S  S S   S   I 
id: (002-002_00_0002)
Scores: (#C #S #D #I) 3 4 1 2
REF:  A M e * N i * N ț i 
HYP:  * N e M A i R E C  i 
Eval: D S   I S   I S S    

With these results:
SPKR | # Snt # Wrd  | Corr    Sub    Del    Ins |   Err  S.Err |
|------+--------------+-------------------------------------|
| 002  |  100    890  |  287    458    145    336   |  939    100

I tried modifying accum-grad: 1 (as I have 1 GPU), grad-clip: 2 and keeping lr:10.0 and the graphs are smoother but the CTC loss explodes again.

![gradclip2](https://user-images.githubusercontent.com/34603830/109790065-a6d5fb00-7c19-11eb-9648-3be0fda09de5.png)


I read somewhere that the CTC first learns to output only < blank > and then begins to insert letters too, is that the case? Should I leave the model to train for longer? Thank you!

**Comment by sw005320** [comment]
Thanks for your report.
Year, can you train a model longer?
The convergence of the voxforge Italian recipe is slow (maybe we need to tune some optimization parameters, but we leave it as it is).

**Comment by tudor-mare** [comment]
Hello!

I have come back with another report. I've left the model to train for 400 epochs and those are the results:

![main](https://user-images.githubusercontent.com/34603830/110306289-56d6aa00-8006-11eb-81fc-0307778909ac.png)

![validation](https://user-images.githubusercontent.com/34603830/110306325-60601200-8006-11eb-8fce-cc412777fcd6.png)
I thought maybe the CTC CER will decrease and it did but it hit a plateau and doesn't seem to move from there. Also the plateau is around 0.95 CER so I have don't think it will go down pretty soon.

I also took a look at what the Encoder Attention is doing:
The first epoch gives an attention like this but after some epochs output is only vertical lines. From one epoch to another there are very big changes, the line move from left to right side very drastically. Should I decrease LR? 
![epoch1](https://user-images.githubusercontent.com/34603830/110308995-7a4f2400-8009-11eb-9dd9-f5c930265504.png)

Next plot is for epoch 444. From what I understand this mask should be diagonal.
![epoch444](https://user-images.githubusercontent.com/34603830/110306713-c6e53000-8006-11eb-91e1-17ce42f94521.png)


I am thinking of training this model with SGD and maybe a smaller learning rate. Any help is highly appreciated!

**Comment by sw005320** [comment]
Are you working on your own data?
If so, it would be better to check the data preparation or data themselves.
I created several recipes from scratch and without any tuning of some existing configs, but it's usually working if we have more than 100 hours of data.

**Comment by tudor-mare** [comment]
Thank you very much!
I took another look at my data preparation and you were right, I mismanaged my labels. Now the training looks like this:
![image](https://user-images.githubusercontent.com/34603830/111283830-3337e300-8648-11eb-905e-a0eebc85f217.png)
![image](https://user-images.githubusercontent.com/34603830/111283874-3c28b480-8648-11eb-9914-3b033a82d7e3.png)
Now I am going to look into further improving accuracy. Thank you a lot!
